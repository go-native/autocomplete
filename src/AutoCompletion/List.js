import React, { Component } from 'react';

class List extends Component {
    render() {
        return (
            <ul className="list-group">
                {this.props.children}
            </ul>
        );
    }
}

export default List;