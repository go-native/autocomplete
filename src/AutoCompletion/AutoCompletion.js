import React, { Component } from 'react';
import './AutoCompletion.css';
import List from './List';
import Item from './Item';

class AutoCompletion extends Component {
  constructor(props) {
    super(props);
    this.state = { showList: true, value: "", activeIndex: null };
    this.handleChange = this.handleChange.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.selectItem = this.selectItem.bind(this);
  }
  handleBlur(e) {
    this.setState({
      showList: false
    });
  }
  handleFocus(e) {
    this.setState({
      showList: true,
    });
  }
  handleChange(e) {
    this.setState({
      value: e.target.value,
      showList: true,
      activeIndex: null
    })
    this.props.onChange(e.target.value);
  }
  selectItem(item) {
    this.props.onSelect(item);
    this.setState({
      value: item.name,
      showList: false,
      activeIndex: null
    });    
  }
  render() {
    return (
      <form className="autocomplete-cnt" tabIndex="0" onBlur={this.handleBlur}>
        <div className="form-group">
          <input type="text" value={this.state.value} onFocus={this.handleFocus} className="form-control" placeholder="Search here..." onChange={this.handleChange} />
        </div>
        {this.state.showList && <List>
          {this.props.items.map((i, index) => <Item key={i.id} onSelect={() => this.selectItem(i)} caption={i.name} active={this.props.items[this.state.activeIndex] === i} onActive={() => this.setState({ activeIndex: index })} />)}
        </List>}
      </form>
    );
  }
}

export default AutoCompletion;