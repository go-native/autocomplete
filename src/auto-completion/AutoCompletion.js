import React, { Component } from 'react';
import { debounce } from 'throttle-debounce';
import List from './List';
import Item from './Item';

const DEBOUNCE_TIMEOUT = 500;
class AutoCompletion extends Component {
  constructor(props) {
    super(props);
    this.state = { showList: true, value: "", activeIndex: null };
    this.handleChange = this.handleChange.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.selectItem = this.selectItem.bind(this);
    this.debounceChange = debounce(DEBOUNCE_TIMEOUT, this.debounceChange);
  }
  handleBlur(e) {
    this.setState({
      showList: false
    });
  }
  handleFocus(e) {
    this.setState({
      showList: true,
    });
  }
  handleChange(e) {
    this.setState({
        value: e.target.value,
        showList: true,
        activeIndex: null
    })
    this.debounceChange(e.target.value);
  }
  debounceChange(value) {
    this.props.onChange(value);
  }
  selectItem(item) {
    this.props.onSelect(item);
    this.setState({
      value: item.name,
      showList: false,
      activeIndex: null
    });    
  }
  render() {
    return (
      <form className="autocomplete-cnt" tabIndex="0" onBlur={this.handleBlur}>
        <div className="form-group">
          <input type="text" value={this.state.value} onFocus={this.handleFocus} className="form-control" placeholder="Search here..." onChange={this.handleChange} />
          {this.props.isFetching && <label>Fetching</label>}
        </div>
        {this.state.showList && <List>
          {this.props.items.map((i, index) => <Item key={i.id} onSelect={() => this.selectItem(i)} caption={i.name} active={this.props.items[this.state.activeIndex] === i} onActive={() => this.setState({ activeIndex: index })} />)}
        </List>}
      </form>
    );
  }
}

export default AutoCompletion;
