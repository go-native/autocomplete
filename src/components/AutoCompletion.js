import React, { Component } from 'react';
import { debounce } from 'throttle-debounce';
import List from './List';
import Item from './Item';

const DEBOUNCE_TIMEOUT = 500;
class AutoCompletion extends Component {
  constructor(props) {
    super(props);
    this.state = { showList: true, value: "", activeIndex: null };
    this.handleChange = this.handleChange.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.selectItem = this.selectItem.bind(this);
    this.debounceChange = debounce(DEBOUNCE_TIMEOUT, this.debounceChange);
    this.arrowMove = this.arrowMove.bind(this);
  }
  handleBlur(e) {
    this.setState({
      showList: false
    });
  }
  handleFocus(e) {
    this.setState({
      showList: true,
    });
  }
  handleChange(e) {
    this.setState({
        value: e.target.value,
        showList: true,
        activeIndex: null
    })
    // Debounce the function call to avoid multiple requests
    this.debounceChange(e.target.value);
  }
  debounceChange(value) {
    this.props.onChange(value);
  }
  selectItem(item) {
    // Propogate the item to the parent
    this.props.onSelect(item);
    this.setState({
      value: item.name,
      showList: false,
      activeIndex: null
    });    
  }
  arrowMove(e) {
    if (e.key === 'ArrowDown') {
        // Looping through the items from top to bottom
        e.preventDefault();
        let activeIndex = (this.state.activeIndex === null) ? -1 : this.state.activeIndex;
        let index = (((activeIndex + 1) % this.props.items.length) + this.props.items.length) % this.props.items.length;
        this.setState({
            activeIndex: index
        });
    }
    if (e.key === 'ArrowUp') {
        // Looping through the items from bottom to top
        e.preventDefault();
        let activeIndex = (this.state.activeIndex === null) ? 10 : this.state.activeIndex;
        let index = (((activeIndex - 1) % this.props.items.length) + this.props.items.length) % this.props.items.length;
        this.setState({
            activeIndex: index
        });
    }
    if (e.key === 'Enter') {
        // Select an item when the Enter key is hit
        e.preventDefault();
        this.selectItem(this.props.items[this.state.activeIndex]);
    }   
  }
  render() {
    return (
      <form className="autocomplete-cnt" tabIndex="0" onBlur={this.handleBlur}>
        <div className="form-group">
          <input type="text" 
            onKeyDown={this.arrowMove} 
            value={this.state.value} 
            onFocus={this.handleFocus} 
            className="form-control" 
            placeholder="Search here..." 
            onChange={this.handleChange} 
          />
          {this.props.isFetching && <label>Fetching</label>}
        </div>
        {this.state.showList && <List>
          {this.props.items.map((i, index) => <Item key={i.id} onSelect={() => this.selectItem(i)} caption={i.name} active={this.props.items[this.state.activeIndex] === i} onActive={() => this.setState({ activeIndex: index })} />)}
        </List>}
      </form>
    );
  }
}

export default AutoCompletion;
