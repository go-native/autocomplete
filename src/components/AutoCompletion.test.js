import React from 'react';
import renderer from 'react-test-renderer';
import AutoCompletion from './AutoCompletion';

const ALL_REPOS_MOCK = [
  { id: 1, name: "Tensorflow" },
  { id: 3, name: "VisualStudioCode" },
  { id: 2, name: "Quandoo" }
]
function searchReposMock(query) {
  this.setState({ repos: ALL_REPOS.filter(r => r.name.indexOf(query) !== -1) })
}
describe('<AutoCompletion />', () => {
  it('renders', () => {
    const tree = renderer
      .create(
        <AutoCompletion
          items={ALL_REPOS_MOCK}
          onSelect={(selectedItem) => console.log(selectedItem)}
          onChange={(query) => searchReposMock((query))}
        />,
    )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
