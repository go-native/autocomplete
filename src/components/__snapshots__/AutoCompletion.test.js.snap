// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoCompletion /> renders 1`] = `
<form
  className="autocomplete-cnt"
  onBlur={[Function]}
  tabIndex="0"
>
  <div
    className="form-group"
  >
    <input
      className="form-control"
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="Search here..."
      type="text"
      value=""
    />
  </div>
  <ul
    className="list-group"
  >
    <li
      className="list-group-item"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      Tensorflow
    </li>
    <li
      className="list-group-item"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      VisualStudioCode
    </li>
    <li
      className="list-group-item"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      Quandoo
    </li>
  </ul>
</form>
`;
